File: cli/internal/shareurl/shareurl.go
// Package shareurl builds and parses ooshare share URLs.
//
//	https://ooshare.io/s/<alias|uuid>?lng=en#<base64url master key>
package shareurl

import (
	"errors"
	"fmt"
	"net/url"
	"strings"

	"github.com/dhdtech/only-once-share/cli/internal/crypto"
)

const keyBytes = 32

// ShareURL is a parsed share URL.
type ShareURL struct {
	ID   string // path id: 8-char alias or UUID
	Lang string // ?lng= parameter (empty when absent)
	Key  []byte // master key (decoded from the fragment)
}

// Build constructs the share URL. lang defaults to "en" when empty.
func Build(origin, id, lang, keyB64url string) string {
	origin = strings.TrimRight(origin, "/")
	if lang == "" {
		lang = "en"
	}
	return fmt.Sprintf("%s/s/%s?lng=%s#%s", origin, id, lang, keyB64url)
}

// Parse decodes a full share URL.
func Parse(raw string) (*ShareURL, error) {
	u, err := url.Parse(raw)
	if err != nil {
		return nil, fmt.Errorf("invalid URL: %w", err)
	}
	path := strings.TrimPrefix(u.Path, "/")
	id := strings.TrimPrefix(path, "s/")
	if id == "" || id == path {
		return nil, errors.New("missing secret id in URL path")
	}
	if u.Fragment == "" {
		return nil, errors.New("missing master key in URL fragment")
	}
	key, err := crypto.Base64urlDecode(u.Fragment)
	if err != nil {
		return nil, errors.New("invalid master key in URL fragment")
	}
	if len(key) != keyBytes {
		return nil, fmt.Errorf("invalid master key length %d (want 32)", len(key))
	}
	return &ShareURL{ID: id, Lang: u.Query().Get("lng"), Key: key}, nil
}

File: extension/src/lib/shareurl.ts
const KEY_BYTES = 32;

function isOoshareHost(hostname: string): boolean {
  return (
    hostname === "ooshare.io" ||
    hostname === "www.ooshare.io" ||
    hostname.endsWith(".ooshare.io")
  );
}

export interface ParsedShareUrl {
  id: string;
  lang: string;
  /** base64url-encoded 32-byte master key, exactly as it appears in the fragment */
  key: string;
}

function decodeB64url(s: string): Uint8Array {
  const b64 = s.replace(/-/g, "+").replace(/_/g, "/");
  const padded = b64 + "=".repeat((4 - (b64.length % 4)) % 4);
  const bin = atob(padded);
  const out = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
  return out;
}

export function buildShareUrl(
  origin: string,
  id: string,
  lang: string,
  keyB64url: string,
): string {
  const base = origin.replace(/\/+$/, "");
  const lng = lang || "en";
  return `${base}/s/${id}?lng=${lng}#${keyB64url}`;
}

export function parseShareUrl(raw: string): ParsedShareUrl {
  let u: URL;
  try {
    u = new URL(raw);
  } catch {
    throw new Error("Invalid URL");
  }
  if (!isOoshareHost(u.hostname)) {
    throw new Error("Missing secret id in URL path");
  }
  const path = u.pathname.replace(/^\/+/, "");
  const id = path.startsWith("s/") ? path.slice(2) : path;
  if (!id) throw new Error("Missing secret id in URL path");
  if (!u.hash) throw new Error("Missing master key in URL fragment");
  const key = u.hash.slice(1);
  if (decodeB64url(key).length !== KEY_BYTES) {
    throw new Error("Invalid master key in URL fragment");
  }
  return { id, lang: u.searchParams.get("lng") ?? "", key };
}
